File: packages/ui-swiftui/Sources/KinetixUI/Dialog.swift
//
// Dialog.swift — KinetixDialog family.
//
// Mirrors packages/ui/src/components/dialog.tsx. Rather than SwiftUI's
// `.sheet` (a bottom card on iOS), this is a centred modal: place it in
// an `.overlay { }` or a top-level `ZStack` and drive it with an
// `isPresented` binding (Radix's controlled `open` shape, no `Trigger` /
// `Portal` graph to reassemble — same call as the Compose port wrapping
// `androidx…Dialog`). `max-w-lg` (448) is off the shared spacing scale.
// The backdrop blur has no cheap SwiftUI equivalent and isn't
// approximated — the dim scrim alone reads as modal.
//
// `dismissible: false` (used by KinetixAlertDialog) drops the close
// affordance and scrim-tap dismissal — "must choose an explicit action".
//

import SwiftUI

public struct KinetixDialog<Content: View>: View {
    @Environment(\.kinetixColors) private var colors

    @Binding private var isPresented: Bool
    private let dismissible: Bool
    private let content: Content

    public init(
        isPresented: Binding<Bool>,
        dismissible: Bool = true,
        @ViewBuilder content: () -> Content
    ) {
        self._isPresented = isPresented
        self.dismissible = dismissible
        self.content = content()
    }

    public var body: some View {
        if isPresented {
            ZStack {
                Color.black.opacity(0.4)
                    .ignoresSafeArea()
                    .onTapGesture { if dismissible { isPresented = false } }

                VStack(alignment: .leading, spacing: 16) { content }
                    .padding(24) // p-6
                    .frame(maxWidth: 448) // max-w-lg
                    .background(colors.background, in: RoundedRectangle(cornerRadius: 16, style: .continuous))
                    .overlay(alignment: .topTrailing) {
                        if dismissible {
                            Button { isPresented = false } label: {
                                Image(systemName: "xmark")
                                    .font(.system(size: 14, weight: .medium))
                                    .foregroundStyle(colors.mutedForeground)
                            }
                            .buttonStyle(.plain)
                            .padding(16)
                            .accessibilityLabel("Close")
                        }
                    }
                    .overlay {
                        RoundedRectangle(cornerRadius: 16, style: .continuous)
                            .strokeBorder(colors.border, lineWidth: 1)
                    }
                    .shadow(color: .black.opacity(0.2), radius: 16, y: 8)
                    .padding(24)
            }
            .transition(.opacity)
        }
    }
}

public struct KinetixDialogHeader<Content: View>: View {
    private let content: Content

    public init(@ViewBuilder content: () -> Content) {
        self.content = content()
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 6) { content }
    }
}

public struct KinetixDialogFooter<Content: View>: View {
    private let content: Content

    public init(@ViewBuilder content: () -> Content) {
        self.content = content()
    }

    public var body: some View {
        HStack(spacing: 8) {
            Spacer(minLength: 0)
            content
        }
    }
}

public struct KinetixDialogTitle: View {
    @Environment(\.kinetixColors) private var colors
    private let text: String

    public init(_ text: String) {
        self.text = text
    }

    public var body: some View {
        Text(text)
            .font(.system(size: 18, weight: .semibold))
            .tracking(-0.4) // tracking-tight
            .foregroundStyle(colors.foreground)
    }
}

public struct KinetixDialogDescription: View {
    @Environment(\.kinetixColors) private var colors
    private let text: String

    public init(_ text: String) {
        self.text = text
    }

    public var body: some View {
        Text(text)
            .font(.system(size: 14))
            .foregroundStyle(colors.mutedForeground)
            .fixedSize(horizontal: false, vertical: true)
    }
}

File: packages/ui-compose/ui/src/main/kotlin/com/kinetixui/ui/Dialog.kt
package com.kinetixui.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.window.Dialog

/**
 * KinetixDialog family — mirrors `packages/ui/src/components/dialog.tsx`.
 * Radix's `Dialog`/`DialogPortal`/`DialogOverlay`/`DialogTrigger`/
 * `DialogClose` sub-components aren't reproduced 1:1: this wraps Compose's
 * own `androidx.compose.ui.window.Dialog`, which already owns a separate
 * window, a dim scrim, and back-press/outside-tap dismissal — the same
 * "reuse the platform's overlay machinery" call `KinetixSlider` made for
 * drag gestures. The caller supplies `visible`/`onDismissRequest` directly
 * (Radix's own controlled `open`/`onOpenChange` shape, just without a
 * separate `Portal`/`Root` graph to reassemble) rather than a `Trigger`
 * sub-component. `max-w-lg` (448dp) isn't on the shared `spacing_*` scale
 * — hardcoded, same reasoning as `KinetixFab`'s off-scale sizes. The
 * backdrop blur has no direct Compose equivalent and isn't approximated —
 * the dim scrim alone reads as "modal" without it.
 */
@Composable
fun KinetixDialog(
    visible: Boolean,
    onDismissRequest: () -> Unit,
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit,
) {
    if (!visible) return
    val colors = KinetixColorScheme.current
    val shape = RoundedCornerShape(dimensionResource(R.dimen.radius_xl))

    Dialog(onDismissRequest = onDismissRequest) {
        Box(
            modifier = modifier
                .widthIn(max = 448.dp) // max-w-lg, not on the shared scale
                .clip(shape)
                .background(colors.background, shape)
                .border(dimensionResource(R.dimen.border_width_default), colors.border, shape),
        ) {
            Column(
                modifier = Modifier.padding(dimensionResource(R.dimen.spacing_6)),
                verticalArrangement = Arrangement.spacedBy(dimensionResource(R.dimen.spacing_4)),
            ) {
                CompositionLocalProvider(LocalContentColor provides colors.foreground) {
                    content()
                }
            }
            Text(
                text = "×",
                color = colors.mutedForeground,
                modifier = Modifier
                    .align(Alignment.TopEnd)
                    .padding(dimensionResource(R.dimen.spacing_4))
                    .clickable(onClick = onDismissRequest),
            )
        }
    }
}

@Composable
fun KinetixDialogHeader(
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit,
) {
    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(6.dp)) { // space-y-1.5, not on the shared scale
        content()
    }
}

@Composable
fun KinetixDialogFooter(
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit,
) {
    Row(
        modifier = modifier,
        horizontalArrangement = Arrangement.spacedBy(dimensionResource(R.dimen.spacing_2), Alignment.End),
    ) {
        content()
    }
}

@Composable
fun KinetixDialogTitle(
    text: String,
    modifier: Modifier = Modifier,
) {
    val fontSize = dimensionResource(R.dimen.font_size_title_dialog).value.sp // text-lg (18dp)
    Text(
        text = text,
        modifier = modifier,
        fontWeight = FontWeight.SemiBold,
        fontSize = fontSize,
        lineHeight = fontSize, // leading-none
        letterSpacing = dimensionResource(R.dimen.letter_spacing_tighter).value.sp,
    )
}

@Composable
fun KinetixDialogDescription(
    text: String,
    modifier: Modifier = Modifier,
) {
    val colors = KinetixColorScheme.current
    Text(
        text = text,
        modifier = modifier,
        color = colors.mutedForeground,
        fontSize = dimensionResource(R.dimen.font_size_body_md).value.sp,
        lineHeight = dimensionResource(R.dimen.line_height_body_md).value.sp,
    )
}
